apps/test/netsim/NetSimRemoteNodeSelectionPanel.js

```diff
@@ -5,25 +5,162 @@ var NetSimTestUtils = require('../util/netsimTestUtils');
+      nodes.push(new NetSimClientNode());
+    }
+    return nodes;
+  }
+
+  function panelWithNodes(clients) {
+    return new NetSimRemoteNodeSelectionPanel(rootDiv, {
+      nodesOnShard: clients,
+      incomingConnectionNodes: []
+    }, emptyCallbacks);
+  }
+
+  beforeEach(function () {
+    NetSimTestUtils.initializeGlobalsToDefaultValues();
+    rootDiv = $(document.createElement('div'));
+    emptyCallbacks = {
+      addRouterCallback: function () {},
+      cancelButtonCallback: function () {},
+      joinButtonCallback: function () {},
+      resetShardCallback: function () {}
+    };
+  });
+
+  describe("canAddRouter", function () {
+    var MAX_ROUTERS;
 
     beforeEach(function () {
-      NetSimTestUtils.initializeGlobalsToDefaultValues();
-      rootDiv = $(document.createElement('div'));
-      panel = new NetSimRemoteNodeSelectionPanel(rootDiv, {
-        nodesOnShard: [],
+      NetSimGlobals.getLevelConfig().showAddRouterButton = true;
+      MAX_ROUTERS = NetSimGlobals.getMaxRouters();
+    });
+
+    it ("true in empty shard (with default test setup)", function () {
+      var panel = panelWithNodes([]);
+      assert(panel.canAddRouter());
+    });
+
+    it ("false if level.showAddRouterButton is false", function () {
+      NetSimGlobals.getLevelConfig().showAddRouterButton = false;
+      var panel = panelWithNodes([]);
+      assert(!panel.canAddRouter());
+    });
+
+    it ("false with an outgoing connection request", function () {
+      var nodes = makeClients(3);
+      var panel = new NetSimRemoteNodeSelectionPanel(rootDiv, {
+        nodesOnShard: nodes,
+        remoteNode: nodes[0],
         incomingConnectionNodes: []
-      }, {
-        addRouterCallback: function () {},
-        cancelButtonCallback: function () {},
-        joinButtonCallback: function () {},
-        resetShardCallback: function () {}
+      }, emptyCallbacks);
+      assert(!panel.canAddRouter());
+    });
+
+    // A single-part address imposes no limit on the number of addressable
+    // routers, so the global limit is used.
+    describe("with address format '4'", function () {
+      beforeEach(function () {
+        NetSimGlobals.getLevelConfig().addressFormat = '4';
+      });
+
+      it ("true if current router count is below the global limit", function () {
+        var panel = panelWithNodes(makeRouters(MAX_ROUTERS - 1));
+        assert(panel.canAddRouter());
+      });
+
+      it ("false if current router count is at/beyond the global limit", function () {
+        var panel = panelWithNodes(makeRouters(MAX_ROUTERS));
+        assert(!panel.canAddRouter());
+      });
+
+      it ("true if current client count is at/beyond the global router limit", function () {
+        var panel = panelWithNodes(makeClients(MAX_ROUTERS));
+        assert(panel.canAddRouter());
+      });
+
+    });
+
+    // The two-bit router part imposes a limit of four addressable routers.
+    describe ("with address format '2.8'", function () {
```

Comment:
should this be `"with address format '2.2'"`? Likewise the one below. If not, I'm confused about something.